packet_link_qualification/index.md

```diff
@@ -0,0 +1,734 @@
+// will be responsible for both setting the interface into a loopback as well
+// as injecting individual packets up to packet_count into the closed loop at
+// the packet_size. These packets will form a closed loop as both sides of the
+// loop will forward the same set of packets for the duration of the
+// qualification.
+message PacketInjectorConfiguration {
+  // Number of packets to inject into the closed loop.
+  uint32 packet_count = 1;
+
+  // Size of packets to inject.
+  // if unspecified, the default value is 1500 bytes.
+  uint32 packet_size = 2;
+
+  // Loopback mode for this qualification.
+  oneof loopback_mode  {
+    // PMD based loopback encompass either PHY based port
+    // loopbacks or port based ASIC loopbacks which do
+    // use forwarding engine processing.
+    // Their use may limit the stats available for the
+    // qualification.
+    PmdLoopbackConfiguration pmd_loopback = 101;
+    // ASIC based loops are done inside the forwarding
+    // engine and must have stats available to the
+    // qualification.
+    AsicLoopbackConfiguration asic_loopback = 102;    
+  }
+}
+
+
+message PmdLoopbackConfiguration {
+}
+
+message AsicLoopbackConfiguration {
+  // This is where any l2/l3/l4 match criteria would be described.
+}
+```
+
+#### Get Messages
+
+```
+// GetRequest returns the status for the provided ids.
+message GetRequest{
+  repeated string ids =1;
+}
+
+// GetResponse returns a map containing the values for all requested
+// Qualification results. If the QualificationResult state is
+// QUALIFICATION_STATE_ERROR the caller should inspect the status field for
+// the exact error code and message.
+// Expected errors codes:
+// NOT_FOUND when the requested id was not found by the service.
+// INVALID_ARGUMENT for any configuration parameter which is unsupported.
+message GetResponse{
+  map<string, QualificationResult> results = 1;
+}
+
+// States of qualification.
+enum QualificationState {
+  QUALIFICATION_STATE_UNSPECIFIED = 0;
+  QUALIFICATION_STATE_ERROR = 1;  // The qualification has errored.
+  QUALIFICATION_STATE_IDLE = 2;       // Initial state for the qualification.
+  QUALIFICATION_STATE_SETUP = 3;       // Interface is being configured.
+  QUALIFICATION_STATE_RUNNING = 4;    // Qualification underway.
+  QUALIFICATION_STATE_TEARDOWN = 5;    // Interface is being reset.
+  QUALIFICATION_STATE_COMPLETED = 6;  // Qualification is complete.
+}
+
+message QualificationResult {
+  // ID of the qualification.
+  string id = 1;
+
+  // Interface name of the qualification.
+  string interface_name = 2;
+
+  // The state the test was in when the results were snapshotted.
+  QualificationState state = 3;
+
+  // The number of qualification packets sent.
+  uint64 packets_sent = 4;
+
+  // The number of qualification packets received.
+  uint64 packets_received = 5;
+
+  // The number of packets transmitted that experienced corruption.
```

Comment:
To make this clearer, we could map these counts to the equivalent counters in the models, such that we don't need to define any semantics for them, however, it wasn't immediately clear that for some of the modes that the counters will be completely equivalent. WDYT?